crates/ruff_linter/src/message/text.rs

```diff
@@ -286,38 +283,61 @@ impl Display for MessageCodeFrame<'_> {
     }
 }
 
-fn replace_whitespace(source: &str, annotation_range: TextRange) -> SourceCode {
+/// Given some source code and an annotation range, this routine replaces
+/// tabs with ASCII whitespace, and unprintable characters with printable
+/// representations of them.
+///
+/// The source code returned has an annotation that is updated to reflect
+/// changes made to the source code (if any).
+fn replace_whitespace_and_unprintable(source: &str, annotation_range: TextRange) -> SourceCode {
     let mut result = String::new();
     let mut last_end = 0;
     let mut range = annotation_range;
     let mut line_width = LineWidthBuilder::new(IndentWidth::default());
 
+    let mut update_range = |index, len| {
+        if index < usize::from(annotation_range.start()) {
+            range += TextSize::new(len - 1);
+        } else if index < usize::from(annotation_range.end()) {
+            range = range.add_end(TextSize::new(len - 1));
+        }
+    };
```

Comment:
This function could use a comment explaining what it does or a more descriptive name. (I have been staring at it for a while trying to make sense of it and I can't quite, and I'm not sure if that's because I see a bug or misunderstand what it's doing)

Reply:
Yes! I added this:

```rust
    // Updates the range given by the caller whenever a single byte (at
    // `index` in `source`) is replaced with `len` bytes.
    //
    // When the index occurs before the start of the range, the range is
    // offset by `len`. When the range occurs after or at the start but before
    // the end, then the end of the range only is offset by `len`.
    let mut update_range = |index, len| {
        if index < usize::from(annotation_range.start()) {
            range += TextSize::new(len - 1);
        } else if index < usize::from(annotation_range.end()) {
            range = range.add_end(TextSize::new(len - 1));
        }
    };
```

(The actual logic here is unchanged. I just moved it to a helper function so I could reuse it.)

Reply:
Ohh ok that's why this was so baffling. In spite of the clear text of the match I couldn't get it out of my head that "obviously" we're replacing "weird unicodey characters" with "plain ascii characters" and therefore this code should be replacing `len` bytes with 1 byte (which it clearly isn't doing).